Geek-Collectors-Network/geek-collectors-network
# Connect frontend to backend

### Description
Whheww! This is a big PR...

#### Docker
 - Remove frontend dockerfiles, services, and any supporting files
 - Add built frontend files to api dockerfile
 - Rename api to app [in docker/docker-compose]

#### Backend
 - Extend catch-all route to render web-app to users
 - Add Helmet CSP to prevent HTTPS redirects on localhost
 - Add WEB_ROOT environment variable

#### Frontend
 - **BREAKING CHANGE:** Remove start:dev script -- Frontend requires an API server to work, you might as well launch the API to access the frontend
 - Makes fetch and mutation api requests for the Registration, Login, Profile, and Account Info pages

### Retrospective
CORs is a pain in the butt because cookies from the api can not be saved to the browser.

`express-session` is not a good solution for client-side rendered frontends.

### Testing Notes
The commands in the Makefile were not designed for interdependence between apps. This means that we'll need to manually run some commands that we would not have before.

This will be solved with the new project management script (not yet uploaded).

### Testing
1. Build the latest frontend: `npm run -w apps/frontend build`
2. Spin up the db: `make res`
3. Add `WEB_ROOT=../frontend/dist` to apps/api/.env
5. Start the api server: `npm run -w apps/api start:dev`
6. Navigate to: http://127.0.0.1:3000/
7. Register for a new account
8. Login with your new account
9. Verify you can see your profile information: http://127.0.0.1:3000/profile/edit
10. Verify you can see your account information: http://127.0.0.1:3000/account
11. In the account information page, edit your first name, then save
12. Refresh the page, verify your change persists
